### DISCUSSION TITLE
Data format to metaprompts for RAG usecase

### DISCUSSION THREAD

**Discussion by Nithyashree-coder** [microsoft/sammo#60] [Help] [answered]
I am working with RAG use case by using [rag_tuning ](https://github.com/microsoft/sammo/blob/main/examples/paper_rag/rag_tuning_sammo.py) code. Can you please check if the way I am passing data context, questions and answers are proper or not? Because there is no documentation on RAG use case on how the RAG prompts are getting optimized. I did check on the aspects to optimise from [this link](https://www.microsoft.com/en-us/research/blog/sammo-a-general-purpose-framework-for-prompt-optimization/). But still it would be really helpful if you give a small example on how data should be framed to be passed in meta prompts. I have tried a sample code here. Can you please check if this is how the data should be passed as per the code I have tried? [SAMMO_RAG](https://github.com/Nithyashree-coder/Sammo_Rag)

**Comment by t-schn** [comment]
Hi, have you looked at the [data & templating tutorial](https://microsoft.github.io/sammo/tutorials/basics_2_working_with_data.html) yet? If so, let me know what's still missing.

**Comment by Nithyashree-coder** [comment]
Hi @t-schn , yes I have read all the documentation. However, it is unclear how I should pass context, question and answer for RAG. 
**Task:**
Embed a list of contexts: context1, context2, context3......contextn and get Embeddings.
I will also have a predefined question and answer in dtrain. 
When a new question arrives, the process should get the similar matching embeddings and rank the most matching embeddings and return the top 5 answers (chunk). I did check on information on EmbeddingFewshotExamples which does this operation on few shot examples. 
**Q1. However is EmbeddingFewshotExamples doing the job of retrieval?**  
**Q2. If not, in the rag_tuning.py example what is the retriever that you are using?**
The example didn't have information on how and what the data is passed. 

I am hardcoding inputs, constant and outputs and converting it into DataTable format. However, I am unclear on how the data is framed.  The way I am passing the data is as follows.

For context (d_context):
constants = None
input = ["context1", "context2", "context3"]
Output = None
<img width="546" alt="image" src="https://github.com/user-attachments/assets/b0a1448a-adaf-4a01-b806-286d9450df81">

For training data: 
d_train:
constants = "Answer the following questions based on the given passage."
input = ["question1", "question2", "question3"]
Output = ["answer1", "answer2", "answer3"]

<img width="716" alt="image" src="https://github.com/user-attachments/assets/a6b552bc-d961-4536-b6d5-4f1b489811f2">


For testing data:
constants = "Answer the following questions based on the given passage."
input = ["question1", "question2", "question3"]
Output = ["answer1", "answer2", "answer3"]

<img width="719" alt="image" src="https://github.com/user-attachments/assets/6be0e58d-a68a-4c09-a7ba-bc7ae758d1da">

I am passing d_context in metaprompts in EmbeddingFewshotExamples and also in input data (InputData). I am also performing instruction optimization to this and I am performing orientation optimization to this. 
**Q3. Can you please let me know if the way I am passing data to metaprompts is correct or not?**
If not please let me know how to pass context, question and answer with an example. 

<img width="411" alt="image" src="https://github.com/user-attachments/assets/bdc784a1-2d55-4edd-9f71-60d8a3f484bc">

**Comment by t-schn** [comment, marked as answer]
@Nithyashree-coder: I'm not sure I 100% understand your questions, so I created a new tutorial that explains RAG with SAMMO in more detail [here](https://github.com/microsoft/sammo/blob/main/docs/tutorials/basics_5_rag.ipynb)

For your questions:

1. EmbeddingFewShotExamples is doing the retrieval, yes. It uses the string representation of the input column in the DataTable. 
2. You pass data via a DataTable. You can use nested records, etc. 
3. This seems correct. 

Hope that helps!

**Reply by Nithyashree-coder** [reply]
Thank you so much. I now got an idea on how the data is being passed.
